clients/roscpp/src/libros/spinner.cpp

```diff
@@ -42,25 +122,24 @@ namespace ros
 
 void SingleThreadedSpinner::spin(CallbackQueue* queue)
 {
-  boost::recursive_mutex::scoped_try_lock spinlock(spinmutex);
-  if(!spinlock.owns_lock()) {
-    ROS_ERROR("SingleThreadedSpinner: You've attempted to call spin "
-              "from multiple threads.  Use a MultiThreadedSpinner instead.");
-    return;
-  }
-
-  ros::WallDuration timeout(0.1f);
-
   if (!queue)
   {
     queue = getGlobalCallbackQueue();
   }
 
+  if (!spinner_monitor.add(queue, true))
+  {
+    ROS_FATAL_STREAM("SingleThreadedSpinner: " << DEFAULT_ERROR_MESSAGE);
+    throw std::runtime_error("There is already another spinner on this queue");
```

Comment:
The original behavior was printing an error message and then returned gracefully ignoring the new request. This behavior should be maintained. Therefore the `ROS_FATAL_STREAM` should be replaced with `ROS_ERROR_STREAM` (since its not fatal anymore, but indicates an error in using the API) and `throw` should be replaced with a `return`.